### PR TITLE
[refactor] Apply forward declaration as much as possible

### PR DESCRIPTION
**Pull request by BewareMyPower**
Fixes https://github.com/apache/pulsar-client-cpp/issues/60

### Motivation

The includes in pulsar-client-cpp is very casual. There are a lot of implicit includes and the forward declaration is not used much. For example, if `lib/ClientConnection.h` was modified, 27 files would be recompiled.

The other problem is the `SortIncludes` attribute in `.clang-format` file is false. It might be okay in early days. However, as the project grows, including many headers without ordering brings a very bad experience. Combining with the very few usages of the forward declaration, it's hard to determine whether a header is still required after a change.

### Modifications

Apply forward declarations as much as possible and change the `SortIncludes` to true in `.clang-format` file.

It's special for the `PulsarApi.pb.h` file because the size of this header is over 1 MiB. For classes we can use forward declaration, but for enumerations we have to include this header.

To solve this problem, `ProtoApiEnums.h` is added to define some constant integers that can be cast implicitly from the enumerations. If we want to use enumerations from `PulsarApi.pb.h`, we can include `ProtoApiEnums.h` instead.

Finally, to unify the include rules, in `lib/*.h`, if we want to include a header (e.g. `xxx.h`) from the same directory, just use `#include "xxx.h"`. Don't use `#include "lib/xxx.h"` or `#include <lib/xxx.h>` in headers of `lib/` directory.

### Improvements

Since forward declaration is applied everywhere now, take `lib/ClientConnection.h` for example. After this patch, only 11 files needs to be recompiled, while 27 files would be recompiled before.

This patch also reduces the binary size and speeds up the compilation time.

Binary size:
- `libpulsar.a`: 319234696 (305 MiB) -> 286716564 (274 MiB)
- `libpulsar.so`: 110162496 (106 MiB) -> 102428456 (98 MiB)

Compilation time with the following commands:

```bash
cmake -B build -DBUILD_TESTS=OFF
cmake --build build -j8
```

From 2m35.640s to 1m44.213s.

### Documentation

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -19,7 +19,7 @@
 BasedOnStyle: Google
 IndentWidth: 4
 ColumnLimit: 110
-SortIncludes: false
+SortIncludes: true
```
Comment (Demogorgon314) [inline .clang-format:22]: The value should be `CaseInsensitive`  or `CaseSensitive `, see https://clang.llvm.org/docs/ClangFormatStyleOptions.html

Comment (BewareMyPower) [reply .clang-format:22]: Currently the `clang-format` binds to 11.0.0, this option is a boolean value, see https://releases.llvm.org/11.0.1/tools/clang/docs/ClangFormatStyleOptions.html

Comment (Demogorgon314) [reply .clang-format:22]: Interesting. Looks like the `clang-format version 11.1.0` changed the option to boolean, and the `clang-format version 14.0.6` changed back. 

```
➜   clang-format -version
clang-format version 14.0.6
➜  clang-format -style=llvm -dump-config | grep SortIncludes
SortIncludes:    CaseSensitive

➜  clang-format-11 -version
clang-format version 11.1.0
➜  clang-format-11 -style=llvm -dump-config | grep SortIncludes
SortIncludes:    true
```

Comment (BewareMyPower) [reply .clang-format:22]: No. This option is a boolean since the `clang-format` 3.9. It's a newer version of clang that changed it.

Comment (BewareMyPower) [reply .clang-format:22]: https://releases.llvm.org/3.9.0/tools/clang/docs/ClangFormatStyleOptions.html

**Comment by BewareMyPower** [comment]
@merlimat @shibd @RobertIndie @Demogorgon314  Please take a look at this PR in prior to other PRs since it brings many changes. Other PRs should be rebased to it.

Though it's a huge PR, most changes are only related to the header includes, which is definitely good for code quality. The code changes are mostly related to the `ProtoApiEnums.h`. Without it, there would still be many files that have to include `PulsarApi.pb.h`.

**Review by merlimat — APPROVED** [review]

**Pull request merged**